src/jit/lsrabuild.cpp

```diff
@@ -1314,92 +1314,96 @@ void LinearScan::buildInternalRegisterUses()
-    // TODO-CQ: Consider reworking this as part of addressing GitHub Issues #18144 and #17481.
-    for (RefInfoListNode *listNode = defList.Begin(), *end = defList.End(); listNode != end;
-         listNode = listNode->Next())
-    {
-        var_types treeType = listNode->treeNode->TypeGet();
-        if (varTypeIsSIMD(treeType) && varTypeNeedsPartialCalleeSave(treeType))
+            continue;
+        }
+        noway_assert(interval.isUpperVector);
+        if (interval.relatedInterval->getVarIndex(compiler) == varIndex)
         {
-            RefPosition* pos = newRefPosition(listNode->ref->getInterval(), currentLoc, RefTypeUpperVectorSaveDef, tree,
-                                              RBM_FLT_CALLEE_SAVED);
+            return &interval;
         }
     }
+    unreached();
 }
 
-// buildUpperVectorRestoreRefPositions - Create special RefPositions for restoring
-//                                       the upper half of a set of large vectors.
+//------------------------------------------------------------------------
+// buildUpperVectorSaveRefPositions - Create special RefPositions for saving
+//                                    the upper half of a set of large vectors.
 //
 // Arguments:
-//    tree             - The current node being handled.
-//    currentLoc       - The location of the current node.
-//    liveLargeVectors - The set of large vector lclVars live across 'tree'.
+//    tree       - The current node being handled
+//    currentLoc - The location of the current node
+//    fpCalleeKillSet - The set of registers killed by this node.
 //
-// Notes:
-//    This is called after all the RefPositions for 'tree' have been created, since the restore,
-//    if needed, will be inserted after the call.
+// Notes: This is called by BuildDefsWithKills for any node that kills registers in the
+//        RBM_FLT_CALLEE_TRASH set. We actually need to find any calls that kill the upper-half
+//        of the callee-save vector registers.
+//        But we will use as a proxy any node that kills floating point registers.
+//        (Note that some calls are masquerading as other nodes at this point so we can't just check for calls.)
 //
-void LinearScan::buildUpperVectorRestoreRefPositions(GenTree*         tree,
-                                                     LsraLocation     currentLoc,
-                                                     VARSET_VALARG_TP liveLargeVectors)
+void LinearScan::buildUpperVectorSaveRefPositions(GenTree* tree, LsraLocation currentLoc, regMaskTP fpCalleeKillSet)
 {
-    assert(enregisterLocalVars);
-    if (!VarSetOps::IsEmpty(compiler, liveLargeVectors))
+    if (enregisterLocalVars && !VarSetOps::IsEmpty(compiler, largeVectorVars))
     {
+        assert((fpCalleeKillSet & RBM_FLT_CALLEE_TRASH) != RBM_NONE);
+        VARSET_TP       liveLargeVectors(VarSetOps::Intersection(compiler, currentLiveVars, largeVectorVars));
```

Comment:
This could use a comment explaining that we will only iterate over the Intersection of Live Vars that are also Large Vectors

Reply:
Will do.